File: 290_Word Pattern_(17th Jan 2022)/WordPattern.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        arr = s.split()
        if len(arr) != len(pattern):
            return False
        
        d = dict()
        
        for i in range(len(pattern)):
            item = d.get(pattern[i])
            if not item:
                if arr[i] in d.values():
                    return False
                d[pattern[i]] = arr[i]
            else:
                if item != arr[i]:
                    return False
            
        return True

File: 290_Word Pattern_(17th Jan 2022)/WordPattern.java
class Solution {
    public boolean wordPattern(String pattern, String s) {
        String[] word=s.split(" ");
        if(word.length!=pattern.length())
            return false;
        HashMap<Character,String> hm=new HashMap<>();
        HashSet<String> hs=new HashSet<>();
        
        for(int i=0;i<pattern.length();i++){
            char ch=pattern.charAt(i);
            if(hm.containsKey(ch)){
                if(!word[i].equals(hm.get(ch)))
                    return false;
            }
            else{
                if(hs.contains(word[i]))
                    return false;
                else{
                hs.add(word[i]);
                hm.put(ch,word[i]);
                }
            }
        }
        return true;
    }
}
